File: Java/src/test/java/com/gildedrose/ItemBuilder.java
package com.gildedrose;

public class ItemBuilder {
    private String name;
    private int sellIn;
    private int quality;

    public ItemBuilder setName(String name) {
        this.name = name;
        return this;
    }

    public ItemBuilder setSellIn(int sellIn) {
        this.sellIn = sellIn;
        return this;
    }

    public ItemBuilder setQuality(int quality) {
        this.quality = quality;
        return this;
    }

    public Item createItem() {
        return new Item(name, sellIn, quality);
    }
}

File: csharpcore/GildedRose/ItemBuilder.cs
using System;

namespace GildedRoseKata;

public class ItemBuilder(string name, int sellIn, int quality) : AbstractItemBuilder(name, sellIn, quality)
{
    public override Item Build()
    {
        if (Quality < ItemQuality.MinQuality || Quality > ItemQuality.MaxQuality)
        {
            throw new ArgumentException("An item's quality must be >=0 and <=50");
        }

        if (ItemType.IsLegendaryItem(Name))
        {
            throw new ArgumentException("Legendary Items cannot be constructed using this builder");
        }

        return base.Build();
    }
}


